Sprint 13 HomeWork

Vozzhd opened the pull request.

Vozzhd committed:
Sprint №13 HomeWork done
Добавлено:
-активити аудиоплеера
-класс аудиоплеера
Изменено:
-класс Track.kt, добавлены поля
-ресурсы

Vozzhd committed:
Sprint №13 HomeWork done
Добавлено:
-переход к треку из истории поиска

Vozzhd committed:
Sprint №13 HomeWork done, ver.3
-вместо передачи разных адаптеров в RecyclerView теперь передаются разные листы
-убран класс HistoryAdapter.kt.
-функция изменения адаптера заменена на изменение листа

faritowich commented on app/src/main/java/com/practicum/playlistmaker/Utilities.kt:
```diff
@@ -3,15 +3,14 @@ package com.practicum.playlistmaker
-    adapter.notifyDataSetChanged()
+fun TrackAdapter.replaceList (list:MutableList<Track>) {
+    this.trackList = list
+    this.notifyDataSetChanged()
```
Можно эту функцию внутри TrackAdapter же объявить, нет особого смысла в экстеншн выносить. Вынос в экстеншны имеет смысл тогда, когда нельзя объявить внутри класса, например, экстеншны для библиотечных классов

Vozzhd replied:
Перенёс, исправил.

faritowich commented on app/src/main/java/com/practicum/playlistmaker/Utilities.kt:
```diff
@@ -3,15 +3,14 @@ package com.practicum.playlistmaker
-fun RecyclerView.replaceAdapter (adapter:  RecyclerView.Adapter<TrackViewHolder>) {
-    this.adapter = adapter
-    adapter.notifyDataSetChanged()
+fun TrackAdapter.replaceList (list:MutableList<Track>) {
```
Для чего `MutableList`? Лучше на `List` заменить

Vozzhd replied:
Везде использовал ранее `MutableList`, решил его же на вход принимать + через эту функцию принимается на вход адаптера лист с историей, который должен быть изменяемым для его опустошения через `clear`.
UPS: Понял-принял, избавился от функции `clear` и `MutableList` правда ведь в любом случае мы будем использовать `MutableList` на уровень выше

faritowich commented on app/src/main/java/com/practicum/playlistmaker/recyclerView/TrackAdapter.kt:
```diff
@@ -28,6 +29,12 @@ class TrackAdapter(
+            val playerActivityIntent = Intent(context,PlayerActivity::class.java)
+            val gson = Gson()
+            val json = gson.toJson(trackList[position])
+            context.startActivity(playerActivityIntent.putExtra("key", json))
```
"key" и вообще строки-ключи стоит выносить в отдельную константу (в данном случае в companion object). Это позволит не бояться забыть поменять в одном месте и забыть в другом. При необходимости просто меняешь в самой константе и всё так же работает. Плюс сразу понятно, где мы передаем и получаем эти данные.

Vozzhd replied:
Понял, принял, вынес

faritowich commented on app/src/main/java/com/practicum/playlistmaker/recyclerView/TrackAdapter.kt:
```diff
@@ -28,6 +29,12 @@ class TrackAdapter(
+            val playerActivityIntent = Intent(context,PlayerActivity::class.java)
+            val gson = Gson()
+            val json = gson.toJson(trackList[position])
+            context.startActivity(playerActivityIntent.putExtra("key", json))
```
Лучше задавать клик лисенеры не в самом адаптере, а вынести их в активити, чтобы вынести из адаптера лишнюю логику, которая не отвечает непосредственно за показ списка. Вариантов реализации несколько, на эту тему можно почитать эту статью: https://antonioleiva.com/recyclerview-listener/

Vozzhd replied:
Почитал, понял принцип, сделал) 
Спасибо огромное, выглядит шикарнейше, работает тоже)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/screens/PlayerActivity.kt:
```diff
@@ -0,0 +1,66 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_player)
+
+        val backButton = findViewById<ImageButton>(R.id.backButton)
```
Рекомендую почитать про библиотеку ViewBinding от гугла и использовать её - она используется повсеместно

Vozzhd replied:
Заменил все `findViewById` на использование `ViewBinding`. Спасибо)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/screens/PlayerActivity.kt:
```diff
@@ -0,0 +1,66 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_player)
+
+        val backButton = findViewById<ImageButton>(R.id.backButton)
```
Рекомендую почитать про библиотеку ViewBinding от гугла и использовать её - она используется повсеместно

Vozzhd replied:
Заменил все `findViewById` на использование `ViewBinding`. Спасибо)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/screens/PlayerActivity.kt:
```diff
@@ -0,0 +1,66 @@
+
+        val gson = Gson()
+        val json = intent.getStringExtra("key")
+        val track = gson.fromJson(json, Track::class.java)
```
Текущая реализация через передачу строки небезопасная, так как если сейчас случайно передать из SearchActivity любой другой класс, то всё будет работать до  момента перехода на этот экран, после чего приложение упадет в рантайме. На самом деле можно передавать сам класс Track. Достаточно сделать его Serializable/Parcelable. Рекомендую почитать про это документацию, обычно делается именно так.

Vozzhd replied:
Разобрался, сделал, спасибо за это)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/screens/PlayerActivity.kt:
```diff
@@ -0,0 +1,66 @@
+        val track = gson.fromJson(json, Track::class.java)
+
+        Glide.with(this)
+            .load(track.artworkUrl100.replaceAfterLast('/', "512x512bb.jpg"))
```
Можно вынести логику по преобразованию ссылки artworkUrl100 в другое разрешение в сам класс Track, создав там соответствующую переменную, например так:
```
    val artworkUrl512 
        get() = artworkUrl100.replaceAfterLast('/', "512x512bb.jpg")
```

Vozzhd replied:
Как вариант можно её сразу при инициализации указать?
`val artworkUrl512: String = artworkUrl100.replaceAfterLast('/', "512x512bb.jpg")`
UPD: Понял ,попробовал, нельзя, сделал как надо, работает. Спасибо)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/screens/PlayerActivity.kt:
```diff
@@ -0,0 +1,66 @@
+        trackNameInPlayer.text = track.trackName
+        artistNameInPlayer.text = track.artistName
+
+        if (!track.trackTimeMillis.isNullOrBlank()) {
```
`if (track.trackTimeMillis.isNotBlank()) `

Vozzhd replied:
Думал лишняя проверка не помешает, исправил)

faritowich commented on app/src/main/java/com/practicum/playlistmaker/recyclerView/TrackAdapter.kt:
```diff
@@ -1,15 +1,16 @@
 import com.practicum.playlistmaker.screens.SearchHistory
 class TrackAdapter(
-    private val trackList: MutableList<Track>,
+    var trackList: MutableList<Track>,
```
Тут сейчас и var, и MutableList. Надо бы оставить что-то одно и менять через него

Vozzhd replied:
Изменил на val, обновление списка сделал не через `reassign` а через `clear` и `addAll`

Vozzhd committed:
Sprint №13 HomeWork done, ver.4
-функция `replaceList` перемещена в TrackAdapter
-вместо типа параметра`MutableList` в `replaceList` теперь принимается `List`
-добавлена переменная внутрь `Track` для обложки в 512х512

Vozzhd committed:
Sprint №13 HomeWork done after review
- применён `bindingView` вместо `findViewById`
- теперь при открытии экрана плеера передаётся на строка в JSON, а класс Track
- вынес `clickListener` в активити из адаптера
